Optionally pass BIP388 policy to signtx

Sjors opened the pull request:
Another step toward #785. This builds on #842.

This adds optional registered BIP388 descriptor policy information to
`signtx`. Existing `signtx` usage remains unchanged when no registration
is supplied.

`signtx` accepts `--registration`, containing the serialized
`RegisteredDescriptor` returned by the `registerdescriptor` command from
#842. The registration contains the policy name, descriptor, device type,
and any device-specific registration data, so separate `--policy-desc`
and `--policy-name` arguments are not needed.

Example:

```bash
hwi --device-type ledger signtx \
  --registration "cmRlc2M..." \
  "cHNidP..."

Sjors committed:
Consistently name sign_tx psbt argument

Sjors commented:
I'll update this after #836, taking the latest changes in #794 into account.

Sjors commented:
Rebased after #836.

Sjors commented:
This now uses the same approach as #791, with the first two commits shared, so I'll keep this draft.

achow101 committed:
descriptor: Track key expression index

For BIP388 policies, we will need to know wht index of each key
expression, so keep track of them explicitly with a key expression index
stored in each PubkeyProvider.

This also mirrors what Bitcoin Core does for descriptors.

achow101 committed:
descriptor: Add functions for making BIP 388 wallet policies

get_bip388_template returns the descriptor as a BIP 388 Wallet
Descriptor Template string.

get_pubkey_providers returns all of the pubkey providers from the
descriptor, in the same order as the placeholders in the bip388
template. Callers can get the strings for the key information vector by
calling PubkeyProvider.to_string_no_deriv()

achow101 committed:
CLI command and boilerplate for registerdescriptor

Adds registerdescriptor CLI command and its handlers.

Adds register_descriptor to HardwareWalletClient with boilerplate
implementation.

achow101 committed:
ledger: Implement register_descriptor

achow101 committed:
bitbox02: Implement register_descriptor

achow101 committed:
Implement register_descriptors for devices that don't support it

BitBox01, Trezor, and Keepkey do not support registering descriptors.
Implement the function as a throw.

achow101 committed:
coldcard: Implement register_descriptor

achow101 committed:
jade: Implement register_descriptor

achow101 committed:
test: Add basic test for register_descriptor

achow101 committed:
descriptor: Don't sort PubkeyProviders in sortedmulti

The sort in sortedmulti occurs only during script expansion. Keep the
pubkeys in the order that they were provided to the parser.

achow101 committed:
descriptors: Actually parse the derivation path

The derivation path should be parsed and validated for correctness, not
just copied directly into a pubkey provider.

achow101 committed:
descriptor: Parse multipath path specifiers

achow101 committed:
descriptor: Add RegisteredDescriptor for holding registration data

After a descriptor is registered, we want to return to the caller some
information about the registration. RegisteredDescriptor is a class that
contains the name, the descriptor, the device type, and any data that
the device returned in response to the registration. This class can be
de/serialized from/to a string.

Sjors committed:
signtx: add BIP388 policy support

Sjors committed:
test: sign with registered descriptor

Extend the generic descriptor registration test to fund, sign, and finalize a transaction spending from the registered policy.

Sjors committed:
ledger: support BIP388 policy signing

Reconstruct the registered wallet policy and pass its registration HMAC when signing the PSBT.

Sjors committed:
coldcard: support BIP388 policy signing

Sign with the existing PSBT flow after registerdescriptor has stored the policy on the device.

Sjors committed:
jade: support BIP388 policy signing

Sign with the existing PSBT flow after registerdescriptor has stored the policy on the device.

Sjors committed:
bitbox02: factor out BIP388 script config

Move descriptor-to-policy translation into a reusable helper without changing registration behavior.

Sjors committed:
bitbox02: support BIP388 policy signing

Translate the registered descriptor into the policy script configuration and use it for transaction inputs and change outputs.

Sjors commented:
Rebased and now using `RegisteredDescriptor` from https://github.com/bitcoin-core/HWI/pull/842. I still need to polish the commits. This also has some overlap with #841 so I suggest reviewing that first.

Sjors committed:
coldcard: iterate simulator multisig paths correctly

HWI's Coldcard simulator patch stores multiple derivation paths per
fingerprint so one seed can represent several multisig cosigners.

get_xfp_paths() iterates fingerprint keys instead of path lists. This
breaks descriptor enrollment and later signing fails with "Unknown
multisig wallet". Iterate over the mapping values instead.

Sjors commented:
ColdCard's multisig simulator patch, needed a patch 12ccec2f3022e51e1014be618cdb7e9864c8220b. This belongs in #842 though.

Sjors commented:
Tidied up the commits a bit. I noticed that our BitBox02 code only allows one occurrence per policy. That's a problem with more complicated recovery paths, but that's perhaps better left for a followup to improve.

Sjors commented:
Rebased after #842. Ready for review, though I suggest prioritizing #841.

Sjors commented:
The two early ledger legacy failures look spurious.

achow101 commented on hwilib/devices/ledger.py:
```diff
@@ -267,6 +281,16 @@ def legacy_sign_tx() -> PSBT:
+                    script_addrtype,
+                    registered_wallet,
+                    registered_descriptor.registration,
+                )
```
In 07833327384f92e3fad68a5f1f187f7e25ee4bb9 "ledger: support BIP388 policy signing"

This should live outside of the loop, there's only one possible wallet policy, it does not need to have the same data set for every input in this loop. Furthermore, if the transaction were to involve both a provided policy and an inferred policy, the inferred would not be signed.

Sjors replied:
This should be better now.

One thing I didn't fix, but could be a good followup, is that if Bitcoin Core adds global xpubs, some multisig policies can be inferred and users have to approve twice.

The tests are now run once with and once without global xpubs.

achow101 commented on hwilib/devices/ledger.py:
```diff
@@ -188,7 +188,11 @@ def get_pubkey_at_path(self, path: str) -> ExtendedKey:
 
     @ledger_exception
-    def sign_tx(self, psbt: PSBT) -> PSBT:
+    def sign_tx(
+        self,
+        psbt: PSBT,
+        registered_descriptor: Optional[RegisteredDescriptor] = None,
+    ) -> PSBT:
```
In 07833327384f92e3fad68a5f1f187f7e25ee4bb9 "ledger: support BIP388 policy signing"

This function signature is being changed without updating any callsites or the base class. That is dangerous.

Sjors replied:
I wanted to avoid the churn of having `raise UnavailableActionError("... does not support BIP388 policy signing")` followed by implementing it. But I've now reordered the commits and done that.

achow101 commented on hwilib/devices/ledger.py:
```diff
@@ -188,7 +188,11 @@ def get_pubkey_at_path(self, path: str) -> ExtendedKey:
+    def sign_tx(
+        self,
+        psbt: PSBT,
+        registered_descriptor: Optional[RegisteredDescriptor] = None,
```
In 07833327384f92e3fad68a5f1f187f7e25ee4bb9 "ledger: support BIP388 policy signing"

I would prefer for this to take a set of descriptors, rather than only one, so that if a transaction involves multiple policies, they can all be signed in one call here.

Sjors replied:
Done, with a test. Not all devices support this, but I cherry-picked 23fb64e7f28e45aa0abb14cd56ba5f8db4c61881 _coldcard: enroll Edge policies with Miniscript_ from #841 so ColdCard edge can handle multiple policies (one approval each).

achow101 commented on test/test_device.py:
```diff
@@ -810,3 +814,28 @@ def test_register_descriptor(self):
+            [],
+            [{self.wpk_rpc.getnewaddress(): 0.5}],
+            0,
+            {"includeWatching": True},
```
In 061b6d61f09b0a41c6be43bd58bf9092b0015cd4 "test: sign with registered descriptor"

`includeWatching` is no longer necessary.

achow101 reviewed (commented):
Needs rebase

570dcb0460ff6c72b2b156f07693922e7847c8e5 "coldcard: iterate simulator multisig paths correctly", 4cd34d20bed11da96103d963fcba9b85127ff7be "bitbox02: factor out BIP388 script config" can be dropped.

Sjors commented:
Rebased after https://github.com/bitcoin-core/HWI/pull/849 and added edge support, which uses a new helpers `PartiallySignedInput.has_signature(fingerprint)` and `has_fingerprint`.
 
Also addressed feedback.

Sjors commented:
> [570dcb0](https://github.com/bitcoin-core/HWI/commit/570dcb0460ff6c72b2b156f07693922e7847c8e5) "coldcard: iterate simulator multisig paths correctly" [...] can be dropped

No it's still needed (or maybe it wasn't, but it definitely is now). Without it all non-edge coldcard jobs fail, e.g. https://github.com/bitcoin-core/HWI/actions/runs/32497281637/job/96820811201?pr=792

achow101 commented on hwilib/devices/ledger.py:
```diff
@@ -234,6 +234,38 @@ def legacy_sign_tx() -> PSBT:
+                descriptor.get_bip388_template(),
+                [p.get_bip388_key_info() for p in descriptor.get_pubkey_providers()],
+            )
+            expanded = descriptor.expand(0)
```
In b947aefe8bab1f7dbe3587f849dbea4380178eff "ledger: support BIP388 policy signing"

I don't think we should use `expand` here as it doesn't play nicely with multipath descriptors. Since that function isn't actually used anywhere else, I'm thinking to drop it. That would also make the musig descriptor implementation easier as we would not have to implement key aggregation.

I think the better way is to have a function on `Descriptor` to return an `Optional[AddressType]` since all of descriptors know their address type.

Sjors replied:
> Since that function isn't actually used anywhere else, I'm thinking to drop it.

Done in `descriptor: remove unused script expansion`.

> function on `Descriptor` to return an `Optional[AddressType]` since all of descriptors know their address type.

Done in `descriptor: add address type helper`.

Sjors commented:
Rebased after #841. Dropped the multisig patch be4d753e36fa6c97bf46c17ebb3b9442a2ecaf1b because it was included in #841 via  86a86c6f26b1fcb8be44b151cead3e3ba661a40f.

Sjors commented:
Pushed an automation fix to dismiss `Transaction signed` and `Account registered`. Noticed while working on #852. They belong here, even though they didn't cause problems.

Sjors commented:
Extracted two more Ledger automations from #794 which handle transactions with longer source / destination addresses (i.e. taproot), not anything MuSig2 specific.

achow101 commented:
ACK 97c13f1e99384816518a48095e59a8a1f9468d7a
